Specifications/QIR/Data-Types.md

```diff
@@ -170,49 +166,36 @@ big integers.
 
 ### Tuples and User-Defined Types
 
-Tuple data, including values of user-defined types, is represented as an
-LLVM structure type.
-The structure type will contain a reference count as a standard header,
-followed by the tuple fields.
+Tuple data, including values of user-defined types, is represented as the corresponding LLVM structure type.
+For instance, a tuple containing two integers, `(Int, Int)`, would be represented in LLVM as `type {i64, i64}`.
 
-Because the `%TupleHeader` type actually appears as part of the structure, LLVM will
-not allow it to be an opaque type.
-Thus, QIR defines an LLVM type for the tuple header:
-
-```LLVM
-%TupleHeader = type { i32 }
-```
-
-For instance, a tuple containing two integers, `(Int, Int)`, would be represented in
-LLVM as `type {%TupleHeader, i64, i64}`.
-
-Tuple handles are simple pointers to the underlying data structure.
-When passed to a callable function, tuples should always be passed by reference,
-even if the source language treats them as immutable.
-
-In some situations, a pointer to a tuple of unknown or variable type may be required.
-To satisfy LLVM's strong typing in such a situation, tuple pointers are passed as
-pointers to the initial tuple header field; that is, as a `%TupleHeader*`.
-These pointers should be cast to pointers to the correct data structures by the
+When passed to a callable function, tuples are passed as a pointer to an opaque LLVM structure, `%Tuple`. The pointer is expected to point to the contained data such that it can be cast to the correct data structures by the
 receiving code.
+This permits to define runtime functions that are common for all tuples.
 For instance, this convention is used for callable wrapper functions; see
 [below](#callable-values-and-wrapper-functions).
 
-Many languages provide immutable tuples, along with operators that allow a modified
-copy of an existing tuple to be created.
-In QIR, this is represented by creating a new copy of the existing tuple and then
-modifying the newly-created tuple in place.
-If the compiler knows that the existing tuple is not used after the modification,
-it is possible to avoid the copy and modify the existing tuple in place.
+Many languages provide immutable tuples, along with operators that allow a modified copy of an existing tuple to be created.
+QIR permits to efficiently support this by requiring the runtime to track and be able to access the following given a `%Tuple*`:
+- The size of the tuple in number of bytes
+- The user count indicating how many handles to the tuple exist in the source code
```

Comment:
I'm thinking "alias" might be a better word than "user".

Reply:
I like the idea! It also makes sense for the default "0" value of the count: there is a user, but no aliases.

Reply:
changed to alias